# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/indexedDB/ActorsParent.cpp
# Commit: 9a6271146478
# Full Hash: 9a6271146478a84cd07ef93c7cf03905889da459
# Author: Simon Giesecke <sgiesecke@mozilla.com>
# Date: 2020-04-22 04:10:25
# Regressor Bug: 1623278
# File Overlap Count: 1
# Description:
#   Bug 1623278 - Eliminate output parameters on GetOrCreateConnection. r=dom-workers-and-storage-reviewers,janv
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D68337
# ==============================================================================

diff -r d65dd19b8a16 -r 9a6271146478 dom/indexedDB/ActorsParent.cpp
--- a/dom/indexedDB/ActorsParent.cpp	Tue Apr 21 14:32:51 2020 +0000
+++ b/dom/indexedDB/ActorsParent.cpp	Tue Apr 21 14:33:06 2020 +0000
@@ -5175,8 +5175,8 @@
     NS_ASSERT_OWNINGTHREAD(ConnectionPool);
   }
 
-  nsresult GetOrCreateConnection(const Database& aDatabase,
-                                 RefPtr<DatabaseConnection>* aConnection);
+  Result<RefPtr<DatabaseConnection>, nsresult> GetOrCreateConnection(
+      const Database& aDatabase);
 
   uint64_t Start(const nsID& aBackgroundChildLoggingId,
                  const nsACString& aDatabaseId, int64_t aLoggingSerialNumber,
@@ -12108,11 +12108,10 @@
   self->AdjustIdleTimer();
 }
 
-nsresult ConnectionPool::GetOrCreateConnection(
-    const Database& aDatabase, RefPtr<DatabaseConnection>* aConnection) {
+Result<RefPtr<DatabaseConnection>, nsresult>
+ConnectionPool::GetOrCreateConnection(const Database& aDatabase) {
   MOZ_ASSERT(!NS_IsMainThread());
   MOZ_ASSERT(!IsOnBackgroundThread());
-  MOZ_ASSERT(aConnection);
 
   AUTO_PROFILER_LABEL("ConnectionPool::GetOrCreateConnection", DOM);
 
@@ -12125,41 +12124,41 @@
 
   MOZ_ASSERT(dbInfo);
 
-  RefPtr<DatabaseConnection> connection = dbInfo->mConnection;
-  if (!connection) {
-    MOZ_ASSERT(!dbInfo->mDEBUGConnectionThread);
-
-    nsCOMPtr<mozIStorageConnection> storageConnection;
-    nsresult rv =
-        GetStorageConnection(aDatabase.FilePath(), aDatabase.DirectoryLockId(),
-                             aDatabase.TelemetryId(), &storageConnection);
-    if (NS_WARN_IF(NS_FAILED(rv))) {
-      return rv;
-    }
-
-    connection = new DatabaseConnection(storageConnection,
-                                        aDatabase.GetFileManagerPtr());
-
-    rv = connection->Init();
-    if (NS_WARN_IF(NS_FAILED(rv))) {
-      return rv;
-    }
-
-    dbInfo->mConnection = connection;
-
-    IDB_DEBUG_LOG(("ConnectionPool created connection 0x%p for '%s'",
-                   dbInfo->mConnection.get(),
-                   NS_ConvertUTF16toUTF8(aDatabase.FilePath()).get()));
-
-#ifdef DEBUG
-    dbInfo->mDEBUGConnectionThread = PR_GetCurrentThread();
-#endif
-  }
-
-  dbInfo->AssertIsOnConnectionThread();
-
-  *aConnection = std::move(connection);
-  return NS_OK;
+  if (dbInfo->mConnection) {
+    dbInfo->AssertIsOnConnectionThread();
+
+    return dbInfo->mConnection;
+  }
+
+  MOZ_ASSERT(!dbInfo->mDEBUGConnectionThread);
+
+  nsCOMPtr<mozIStorageConnection> storageConnection;
+  nsresult rv =
+      GetStorageConnection(aDatabase.FilePath(), aDatabase.DirectoryLockId(),
+                           aDatabase.TelemetryId(), &storageConnection);
+  if (NS_WARN_IF(NS_FAILED(rv))) {
+    return Err(rv);
+  }
+
+  RefPtr<DatabaseConnection> connection =
+      new DatabaseConnection(storageConnection, aDatabase.GetFileManagerPtr());
+
+  rv = connection->Init();
+  if (NS_WARN_IF(NS_FAILED(rv))) {
+    return Err(rv);
+  }
+
+  dbInfo->mConnection = connection;
+
+  IDB_DEBUG_LOG(("ConnectionPool created connection 0x%p for '%s'",
+                 dbInfo->mConnection.get(),
+                 NS_ConvertUTF16toUTF8(aDatabase.FilePath()).get()));
+
+#ifdef DEBUG
+  dbInfo->mDEBUGConnectionThread = PR_GetCurrentThread();
+#endif
+
+  return connection;
 }
 
 uint64_t ConnectionPool::Start(
@@ -13840,10 +13839,12 @@
   AUTO_PROFILER_LABEL("Database::EnsureConnection", DOM);
 
   if (!mConnection || !mConnection->HasStorageConnection()) {
-    nsresult rv = gConnectionPool->GetOrCreateConnection(*this, &mConnection);
-    if (NS_WARN_IF(NS_FAILED(rv))) {
-      return rv;
-    }
+    auto res = gConnectionPool->GetOrCreateConnection(*this);
+    if (NS_WARN_IF(res.isErr())) {
+      return res.unwrapErr();
+    }
+
+    mConnection = res.unwrap();
   }
 
   AssertIsOnConnectionThread();
